### PR TITLE
Added Support for Custom Labels and annotations on OpenMetadata Deployment Object

### LINKED ISSUE #69: Add Support for Custom Labels and annotations on OpenMetadata Deployment Object
Ref - https://openmetadata.slack.com/archives/C02B6955S4S/p1661982545124259

### PR DESCRIPTION
**Pull request by Ashok810**
### Describe your changes :

fixes #69 

#
### Type of change :

- [x] Improvement
- [x] New feature
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developer) document.
- [ ] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: charts/openmetadata/values.yaml [hunk 1]
```diff
@@ -5,6 +5,9 @@ replicaCount: 1
 
 # Overrides the openmetadata config file with the help of Environment Variables
 # Below are defaults as per openmetadata-dependencies Helm Chart Values
+labels: {}
+anootations: {}
```
Comment (akash-jain-10) [inline charts/openmetadata/values.yaml:9]: there is a typo here - 
```suggestion
annotations: {}
```

File: charts/openmetadata/values.yaml [hunk 2]
```diff
@@ -5,6 +5,9 @@ replicaCount: 1
...
+componentName: openmetadata
```
Comment (akash-jain-10) [inline charts/openmetadata/values.yaml:10]: This is not required in `values.yaml`!

File: charts/openmetadata/templates/check-db-migrations-job-hook.yaml [hunk 3]
```diff
@@ -7,6 +7,24 @@ metadata:
     "helm.sh/hook": pre-upgrade 
     "helm.sh/hook-weight": "-1" # we use a smaller weight so it's created before the job  
     "helm.sh/hook-delete-policy": hook-succeeded, hook-failed
+    {{- if .Values.annotations }}
+    {{- range $key, $val := fromYaml .Values.annotations }}
+    {{ $key }}: {{ $val | quote }}
+    {{- end}}
+    {{- end}}
+  labels:
+    app.kubernetes.io/name: '{{ include "OpenMetadata.name" .}}'
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/check-db-migrations-job-hook.yaml:16]: I would recommend using [helper class](https://github.com/open-metadata/openmetadata-helm-charts/blob/fd71bcc9eab12eebc13296f7b3f4a3a1f98dd392/charts/openmetadata/templates/_helpers.tpl) to define labels starting with `app.kubernetes.io/*`!

File: charts/openmetadata/templates/db-migrations-hook-cm.yaml [hunk 4]
```diff
@@ -7,7 +7,25 @@ metadata:
     "helm.sh/hook": pre-upgrade 
     "helm.sh/hook-weight": "-10" # we use a smaller weight so it's created before the job  
     "helm.sh/hook-delete-policy": hook-succeeded, hook-failed
+    {{- if .Values.annotations }}
+    {{- range $key, $val := fromYaml .Values.annotations }}
+    {{ $key }}: {{ $val | quote }}
+    {{- end}}
+    {{- end}}
+  labels:
+    app.kubernetes.io/name: '{{ include "OpenMetadata.name" .}}'
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/db-migrations-hook-cm.yaml:16]: I would recommend using [helper class](https://github.com/open-metadata/openmetadata-helm-charts/blob/fd71bcc9eab12eebc13296f7b3f4a3a1f98dd392/charts/openmetadata/templates/_helpers.tpl) to define labels starting with `app.kubernetes.io/*`!

File: charts/openmetadata/templates/deployment.yaml [hunk 5]
```diff
@@ -2,21 +2,52 @@ apiVersion: apps/v1
...
   name: {{ include "OpenMetadata.fullname" . }}
+  annotations:
+    {{- if .Values.annotations }}
+    {{- range $key, $val := fromYaml .Values.annotations }}
+    {{ $key }}: {{ $val | quote }}
+    {{- end}}
+    {{- end}}
   labels:
-    {{- include "OpenMetadata.labels" . | nindent 4 }}
+    app.kubernetes.io/name: '{{ include "OpenMetadata.name" .}}'
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/deployment.yaml:12]: I would recommend using [helper class](https://github.com/open-metadata/openmetadata-helm-charts/blob/fd71bcc9eab12eebc13296f7b3f4a3a1f98dd392/charts/openmetadata/templates/_helpers.tpl) to define labels starting with `app.kubernetes.io/*`!

File: charts/openmetadata/templates/ingress.yaml [hunk 6]
```diff
@@ -17,11 +17,26 @@ kind: Ingress
 metadata:
   name: {{ $fullName }}
   labels:
-    {{- include "OpenMetadata.labels" . | nindent 4 }}
-  {{- with .Values.ingress.annotations }}
+    app.kubernetes.io/name: '{{ include "OpenMetadata.name" .}}'
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/ingress.yaml:20]: I would recommend using [helper class](https://github.com/open-metadata/openmetadata-helm-charts/blob/fd71bcc9eab12eebc13296f7b3f4a3a1f98dd392/charts/openmetadata/templates/_helpers.tpl) to define labels starting with `app.kubernetes.io/*`!

File: charts/openmetadata/templates/service.yaml [hunk 7]
```diff
@@ -3,11 +3,27 @@ kind: Service
 metadata:
   name: {{ include "OpenMetadata.fullname" . }}
   labels:
-    {{- include "OpenMetadata.labels" . | nindent 4 }}
-  {{- with .Values.service.annotations }}
+    app.kubernetes.io/name: '{{ include "OpenMetadata.name" .}}'
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/service.yaml:6]: I would recommend using [helper class](https://github.com/open-metadata/openmetadata-helm-charts/blob/fd71bcc9eab12eebc13296f7b3f4a3a1f98dd392/charts/openmetadata/templates/_helpers.tpl) to define labels starting with `app.kubernetes.io/*`!

File: charts/openmetadata/templates/serviceaccount.yaml [hunk 8]
```diff
@@ -4,9 +4,25 @@ kind: ServiceAccount
 metadata:
   name: {{ include "OpenMetadata.serviceAccountName" . }}
   labels:
-    {{- include "OpenMetadata.labels" . | nindent 4 }}
-  {{- with .Values.serviceAccount.annotations }}
+    app.kubernetes.io/name: '{{ include "OpenMetadata.name" .}}'
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/serviceaccount.yaml:7]: I would recommend using [helper class](https://github.com/open-metadata/openmetadata-helm-charts/blob/fd71bcc9eab12eebc13296f7b3f4a3a1f98dd392/charts/openmetadata/templates/_helpers.tpl) to define labels starting with `app.kubernetes.io/*`!

File: charts/openmetadata/templates/serviceaccount.yaml [hunk 9]
```diff
@@ -4,9 +4,25 @@ kind: ServiceAccount
...
+    app.kubernetes.io/instance: "{{ .Release.Name }}"
+    app.kubernetes.io/component: "openmetadata"
+    {{- if .Values.labels }}
+    {{- range $key, $val := fromYaml .Values.labels }}
+    {{ $key }}: {{ $val | quote }}
+    {{- end}}
+    {{- end}}
+
   annotations:
+    {{- with .Values.serviceAccount.annotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/serviceaccount.yaml:20]: We should only be adding annotations from `.Values.serviceAccount.annotations` if they are available!

File: charts/openmetadata/templates/service.yaml [hunk 10]
```diff
@@ -3,11 +3,27 @@ kind: Service
...
+    app.kubernetes.io/instance: "{{ .Release.Name }}"
+    app.kubernetes.io/component: "openmetadata"
+    {{- if .Values.labels }}
+    {{- range $key, $val := fromYaml .Values.labels }}
+    {{ $key }}: {{ $val | quote }}
+    {{- end}}
+    {{- end}}
+
   annotations:
+    {{- with .Values.service.annotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/service.yaml:19]: We should only be adding `.Values.service.annotations` if they are available!

**Review by akash-jain-10 — CHANGES_REQUESTED** [review]
Left some minor suggestions. Thank you @Ashok810 for the contribution.

File: charts/openmetadata/values.yaml (hunk 1 above)
Comment (Ashok810) [reply charts/openmetadata/values.yaml]: modified annotations spelling

File: charts/openmetadata/values.yaml (hunk 2 above)
Comment (Ashok810) [reply charts/openmetadata/values.yaml]: removed component name from values.yaml

File: charts/openmetadata/templates/check-db-migrations-job-hook.yaml (hunk 3 above)
Comment (Ashok810) [reply charts/openmetadata/templates/check-db-migrations-job-hook.yaml]: used helper class to define labels starting with app.kubernetes.io/*

File: charts/openmetadata/templates/db-migrations-hook-cm.yaml (hunk 4 above)
Comment (Ashok810) [reply charts/openmetadata/templates/db-migrations-hook-cm.yaml]: used helper class to define labels starting with app.kubernetes.io/*

File: charts/openmetadata/templates/deployment.yaml (hunk 5 above)
Comment (Ashok810) [reply charts/openmetadata/templates/deployment.yaml]: used helper class to define labels starting with app.kubernetes.io/*

File: charts/openmetadata/templates/ingress.yaml (hunk 6 above)
Comment (Ashok810) [reply charts/openmetadata/templates/ingress.yaml]: used helper class to define labels starting with app.kubernetes.io/*

File: charts/openmetadata/templates/service.yaml (hunk 7 above)
Comment (Ashok810) [reply charts/openmetadata/templates/service.yaml]: used helper class to define labels starting with app.kubernetes.io/*

File: charts/openmetadata/templates/serviceaccount.yaml (hunk 8 above)
Comment (Ashok810) [reply charts/openmetadata/templates/serviceaccount.yaml]: used helper class to define labels starting with app.kubernetes.io/*

File: charts/openmetadata/templates/service.yaml (hunk 10 above)
Comment (Ashok810) [reply charts/openmetadata/templates/service.yaml:8]: annotations will be only added from .Values.service.annotations if they are available

File: charts/openmetadata/templates/serviceaccount.yaml (hunk 9 above)
Comment (Ashok810) [reply charts/openmetadata/templates/serviceaccount.yaml:9]: annotations will be only added from .Values.serviceAccount.annotations if they are available

File: charts/openmetadata/templates/_helpers.tpl [hunk 11]
```diff
@@ -35,11 +35,17 @@ Common labels
...
 {{- define "OpenMetadata.labels" -}}
 helm.sh/chart: {{ include "OpenMetadata.chart" . }}
+app.kubernetes.io/component: "{{ .Chart.Name }}"
 {{ include "OpenMetadata.selectorLabels" . }}
 {{- if .Chart.AppVersion }}
 app.kubernetes.io/version: {{ .Chart.AppVersion | quote }}
 {{- end }}
 app.kubernetes.io/managed-by: {{ .Release.Service }}
+{{- if .Values.labels }}
+{{- range $key, $val := fromYaml .Values.labels }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/_helpers.tpl:45]: remove the `fromYaml` from this condition. Its failing when I tested locally due to that!

File: charts/openmetadata/templates/check-db-migrations-job-hook.yaml [hunk 12]
```diff
@@ -7,6 +7,13 @@ metadata:
     "helm.sh/hook": pre-upgrade 
     "helm.sh/hook-weight": "-1" # we use a smaller weight so it's created before the job  
     "helm.sh/hook-delete-policy": hook-succeeded, hook-failed
+    {{- if .Values.annotations }}
+    {{- range $key, $val := fromYaml .Values.annotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/check-db-migrations-job-hook.yaml:11]: Remove `fromYaml` from this condition. Helm charts installation and upgrades are failing due to this!

File: charts/openmetadata/templates/check-db-migrations-job-hook.yaml [hunk 13]
```diff
@@ -7,6 +7,13 @@ metadata:
     "helm.sh/hook": pre-upgrade 
     "helm.sh/hook-weight": "-1" # we use a smaller weight so it's created before the job  
     "helm.sh/hook-delete-policy": hook-succeeded, hook-failed
+    {{- if .Values.annotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/check-db-migrations-job-hook.yaml:10]: Let's move this piece of code in `_helpers.tpl` file. and refactor all the files.

```suggestion
    {{- define "OpenMetadata.annotations" -}}
    {{- if .Values.annotations }}
    {{- range $key, $val := .Values.annotations }}
    {{ $key }}: {{ $val | quote }}
    {{- end}}
    {{- end}}
    {{- end }}
```
And then, all the Kubernetes Objects should refer to this by calling
`{{- include "OpenMetadata.annotations" . | nindent 4 }}`

File: charts/openmetadata/templates/db-migrations-hook-cm.yaml [hunk 14]
```diff
@@ -7,7 +7,14 @@ metadata:
     "helm.sh/hook": pre-upgrade 
     "helm.sh/hook-weight": "-10" # we use a smaller weight so it's created before the job  
     "helm.sh/hook-delete-policy": hook-succeeded, hook-failed
+    {{- if .Values.annotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/db-migrations-hook-cm.yaml:10]: All the Kubernetes Objects should refer to this by calling
`{{- include "OpenMetadata.annotations" . | nindent 4 }}`.

Refactor to _helper.tpl as mentioned above in review.

File: charts/openmetadata/templates/deployment.yaml [hunk 15]
```diff
@@ -2,21 +2,31 @@ apiVersion: apps/v1
 kind: Deployment
 metadata:
   name: {{ include "OpenMetadata.fullname" . }}
+  annotations:
+    {{- if .Values.annotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/deployment.yaml:6]: All the Kubernetes Objects should refer to this by calling
`{{- include "OpenMetadata.annotations" . | nindent 4 }}`.

Refactor to _helper.tpl as mentioned above in review.

File: charts/openmetadata/templates/deployment.yaml [hunk 16]
```diff
@@ -2,21 +2,31 @@ apiVersion: apps/v1
...
 spec:
   replicas: {{ .Values.replicaCount }}
   selector:
     matchLabels:
-      {{- include "OpenMetadata.selectorLabels" . | nindent 6 }}
+      {{- include "OpenMetadata.labels" . | nindent 6 }} 
+
   template:
     metadata:
       {{- with .Values.podAnnotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/deployment.yaml:22]: All the Kubernetes Objects should refer to this by calling
`{{- include "OpenMetadata.annotations" . | nindent 8 }}`.

Refactor to _helper.tpl as mentioned above in review.

File: charts/openmetadata/templates/ingress.yaml [hunk 17]
```diff
@@ -17,9 +17,9 @@ kind: Ingress
 metadata:
   name: {{ $fullName }}
   labels:
-    {{- include "OpenMetadata.labels" . | nindent 4 }}
-  {{- with .Values.ingress.annotations }}
+    {{- include "OpenMetadata.labels" . | nindent 4 }} 
   annotations:
+    {{- with .Values.ingress.annotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/ingress.yaml:22]: All the Kubernetes Objects should refer to this by calling
`{{- include "OpenMetadata.annotations" . | nindent 4 }}`.

Refactor to _helper.tpl as mentioned above in review.

Comment (akash-jain-10) [reply charts/openmetadata/templates/ingress.yaml:22]: will keep this condition here as this is for Github Actions datree - 
```
{{- if semverCompare "=<1.13-0" $.Capabilities.KubeVersion.GitVersion }}
datree.skip/K8S_DEPRECATED_APIVERSION_1.16: "Ignore that deprecation in old kubernetes instances"
{{- end }}
```

File: charts/openmetadata/templates/service.yaml [hunk 18]
```diff
@@ -3,11 +3,11 @@ kind: Service
 metadata:
   name: {{ include "OpenMetadata.fullname" . }}
   labels:
-    {{- include "OpenMetadata.labels" . | nindent 4 }}
-  {{- with .Values.service.annotations }}
+    {{- include "OpenMetadata.labels" . | nindent 4 }} 
   annotations:
+    {{- with .Values.service.annotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/service.yaml:8]: All the Kubernetes Objects should refer to this by calling
{{- include "OpenMetadata.annotations" . | nindent 4 }}.

Refactor to _helper.tpl as mentioned above in review.

File: charts/openmetadata/templates/serviceaccount.yaml [hunk 19]
```diff
@@ -4,9 +4,9 @@ kind: ServiceAccount
 metadata:
   name: {{ include "OpenMetadata.serviceAccountName" . }}
   labels:
-    {{- include "OpenMetadata.labels" . | nindent 4 }}
-  {{- with .Values.serviceAccount.annotations }}
+    {{- include "OpenMetadata.labels" . | nindent 4 }} 
   annotations:
+    {{- with .Values.serviceAccount.annotations }}
```
Comment (akash-jain-10) [inline charts/openmetadata/templates/serviceaccount.yaml:9]: All the Kubernetes Objects should refer to this by calling
{{- include "OpenMetadata.annotations" . | nindent 4 }}.

Refactor to _helper.tpl as mentioned above in review.

**Review by akash-jain-10 — CHANGES_REQUESTED** [review]

File: charts/openmetadata/templates/_helpers.tpl (hunk 11 above)
Comment (Ashok810) [reply charts/openmetadata/templates/_helpers.tpl]: removed the fromYaml from this condition

File: charts/openmetadata/templates/check-db-migrations-job-hook.yaml (hunk 13 above)
Comment (Ashok810) [reply charts/openmetadata/templates/check-db-migrations-job-hook.yaml]: added {{- include "OpenMetadata.annotations" . | nindent 4 }}

File: charts/openmetadata/templates/db-migrations-hook-cm.yaml (hunk 14 above)
Comment (Ashok810) [reply charts/openmetadata/templates/db-migrations-hook-cm.yaml]: added {{- include "OpenMetadata.annotations" . | nindent 4 }}

File: charts/openmetadata/templates/deployment.yaml (hunk 16 above)
Comment (Ashok810) [reply charts/openmetadata/templates/deployment.yaml:18]: added {{- include "OpenMetadata.annotations" . | nindent 4 }}

File: charts/openmetadata/templates/ingress.yaml (hunk 17 above)
Comment (Ashok810) [reply charts/openmetadata/templates/ingress.yaml:23]: added {{- include "OpenMetadata.annotations" . | nindent 4 }}

File: charts/openmetadata/templates/service.yaml (hunk 18 above)
Comment (Ashok810) [reply charts/openmetadata/templates/service.yaml:9]: added {{- include "OpenMetadata.annotations" . | nindent 4 }}

File: charts/openmetadata/templates/serviceaccount.yaml (hunk 19 above)
Comment (Ashok810) [reply charts/openmetadata/templates/serviceaccount.yaml:10]: added {{- include "OpenMetadata.annotations" . | nindent 4 }}

File: charts/openmetadata/templates/check-db-migrations-job-hook.yaml (hunk 12 above)
Comment (Ashok810) [reply charts/openmetadata/templates/check-db-migrations-job-hook.yaml]: Removed fromYaml from this condition

File: charts/openmetadata/templates/deployment.yaml (hunk 15 above)
Comment (Ashok810) [reply charts/openmetadata/templates/deployment.yaml]: added {{- include "OpenMetadata.annotations" . | nindent 4 }}

**Pull request closed without merging**